### PR TITLE
New Checker for exhausted iterator in nested loops

### PR DESCRIPTION
**Pull request by ranadheerg**
## Type of Changes

|     | Type                   |
| --- | ---------------------- |
|     | :bug: Bug fix          |
| ✓   | :sparkles: New feature |
|    | :hammer: Refactoring   |
|    | :scroll: Docs          |

## Description

The pull request introduces a new checker, looping-through-iterator (W4801), to detect a specific, high-confidence bug pattern: an iterator defined in an outer scope being exhausted by reuse in a nested loop.

This pattern can lead to silent bugs where an inner loop runs completely on the first iteration of an outer loop and then does nothing on all subsequent iterations. Following the maintainers' advice to start small, this checker is intentionally focused on this nested-loop scenario to ensure high accuracy and avoid false positives.

Refs #2996 

The checker specifically looks for an iterator being consumed in a loop that is nested inside at least one other loop. It correctly handles cases where the iterator is re-defined inside the outer loop (shadowing).

It does not yet detect sequential reuse of an iterator in non-nested loops. This could be a potential enhancement in a future PR.

Running this checker locally against the primer projects successfully identified 4 instances of this bug pattern in established repositories, including Django, music21, and Black, demonstrating its real-world utility.

Closes #XXXX

### CODE REVIEW

**Comment by Pierre-Sassoulas** [comment]
Thank you for working on this. Looking at the primer we need to detect when Stopiteration are caught so we do not raise in this case.

**Comment by jacobtylerwalls** [comment]
The first example from Django is also a false positive given that the outer loop `return`s unconditionally after the first iteration.

**Comment by ranadheerg** [comment]
Thank you for the review. I understand the false positives are due to checker not being smart enough to detect the flow and try catch blocks. I will work on enhancing it to detect the break, return statements for unconditional exits by adding control flow analysis for inner and outer loops.

**Comment by ranadheerg** [comment]
@Pierre-Sassoulas @jacobtylerwalls  
I am enhancing the checker to detect unconditional exits (return, break, raise), which will fix the false positives in the Django and music21 examples.

A question regarding the other Django add_extra example that uses next()

Should we only suppress warnings for next() when a try...except StopIteration block guarantees an exit? My thinking is yes, because the original issue's scope was to prevent silent bugs, whereas an uncaught StopIteration from next() is a runtime error.

Let me know your thoughts.

**Comment by Pierre-Sassoulas** [comment]
> Should we only suppress warnings for next() when a try...except StopIteration block guarantees an exit? My thinking is yes, because the original issue's scope was to prevent silent bugs, whereas an uncaught StopIteration from next() is a runtime error.

We already had similar exploration for https://github.com/pylint-dev/pylint/issues/4725 in https://github.com/pylint-dev/pylint/pull/6987 (lot of false positive at the time made us abandon it, but maybe it's time to revisit)

File: tests/checkers/unittest_iterator_checker.py [hunk 1]
```diff
@@ -0,0 +1,1208 @@
+import astroid
+from astroid import nodes
+
+from pylint import interfaces  # Assuming interfaces is needed by MessageTest or similar
+from pylint.checkers.looping_iterator_checker import RepeatedIteratorLoopChecker
+from pylint.testutils import CheckerTestCase, MessageTest
+
+
+class TestRepeatedIteratorLoopChecker(CheckerTestCase):
+    """Tests for RepeatedIteratorLoopChecker."""
```
Comment (jacobtylerwalls) [inline tests/checkers/unittest_iterator_checker.py]: Please follow the [existing pattern](https://pylint.readthedocs.io/en/stable/development_guide/contributor_guide/tests/writing_test.html#functional-tests) for testing messages.

**Review by jacobtylerwalls — CHANGES_REQUESTED** [review]
It should be possible to reduce the size of this PR. Let's start with the tests.

File: tests/checkers/unittest_iterator_checker.py (hunk 1 above)
Comment (ranadheerg) [reply tests/checkers/unittest_iterator_checker.py]: Hi @jacobtylerwalls, I have updated the checker with comprehensive docstrings including Good and Bad examples to clarify the intent of the message. I've also refined the message 'help' text to explain the stateful nature of iterators. The PR now follows the standard Pylint functional testing pattern, and I've added the necessary Towncrier fragment. Ready for re-review!

**Comment by ranadheerg** [comment]
Hi @jacobtylerwalls and @Pierre-Sassoulas
I am returning to this PR after a bit of a break and would love to continue on this.
I've completed the requested test refactoring and cleanup we discussed. I believe the checker is now in a much tighter state and ready for a fresh review. Please let me know if there are any remaining edge cases you'd like me to address!

File: doc/user_guide/configuration/all-options.rst [hunk 2]
```diff
@@ -8,7 +8,6 @@
 Standard Checkers
 ^^^^^^^^^^^^^^^^^
 
-Remember that only ``tool.pylint`` is required, the section title is not. There are specific notes under each example that cover this.
```
Comment (jacobtylerwalls) [inline doc/user_guide/configuration/all-options.rst]: ?

**Review by jacobtylerwalls — COMMENTED** [review]
What is your view about the false positives in Django and black? The Django example uses nested loops to essentially zip-search in one direction through an iterator. The black example also zip-searches in one direction and raises an error in an else branch.

File: doc/user_guide/configuration/all-options.rst (hunk 2 above)
Comment (Pierre-Sassoulas) [reply doc/user_guide/configuration/all-options.rst]: It's possible that we let in a MR that modified generated code, I had this issue too.

**Comment by ranadheerg** [comment]
Hi @jacobtylerwalls — yes, all three are genuine false positives, each hitting a different gap:

**Ansible (`count`)** — `next(count, None)` is deliberate single-element advancement, not bulk consumption. `visit_call` doesn't distinguish the two. Fix: skip the check when the outer call is `next()`.

**Django (`param_iter`)** — `_has_unconditional_exit` bails immediately when it sees an `if` without an `else`, even if there's a `return` after it that both branches hit. Fix: instead of returning `False`, enqueue two sub-paths (through the `if` body, and skipping it) both continuing with the remaining statements — same logic already used for `if/else`.

**Black (`it`)** — This is a `for/else: raise` pattern, a deliberate fail-fast cursor. The author explicitly handles exhaustion by raising, so this isn't a silent bug at all. The checker doesn't look at `for`-`orelse` today. Fix: suppress if the inner `for` loop's `orelse` contains an unconditional exit, and/or if the loop body has a reachable `break` (cursor pattern).

Does that match your read? If so I'll implement — happy to keep them in separate commits if that's easier to review.

**Comment by ranadheerg** [comment]
Another approach would be to narrow the scope by eliminating the next() call and while loops from the current PR. This would help us avoid more complex control flow cases, leaving us with a cleaner scope - just nested for loops with inner loop having conditional breaks via if and if-else branches.

**Comment by jacobtylerwalls** [comment]
> help us avoid more complex control flow cases, leaving us with a cleaner scope

Without having looked at the code in detail to evaluate this suggestion, this does align with what I was going to suggest: generally, we prefer to keep checks "simple and stupid" to keep from cluttering them with special cases, unless we feel the special cases are really tightly scoped. But since we prefer false negatives to false positives, it's still worth evaluating if we can make the check any more forgiving for these cases.

File: pylint/checkers/looping_iterator_checker.py [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+            confidence=interfaces.HIGH,
+        )
+
+    # --- Helper Method ---
+
+    def _find_ancestor_loop(self, node: nodes.NodeNG) -> nodes.For | None:
+        """Walks up the AST from a node to find the first containing loop."""
+        current: nodes.NodeNG | None = node
+        while current:
+            if isinstance(current, nodes.For):
```
Comment (jacobtylerwalls) [inline pylint/checkers/looping_iterator_checker.py]: I haven't looked in detail yet, but is there a reason to not include `while`?

Comment (ranadheerg) [reply pylint/checkers/looping_iterator_checker.py]: It is intentional to keep the checker simple based on our earlier discussion. While loops always consume iterators with next and it feels like including them generates false positives. 
Restricting both loops to for keeps the checker with highest confidence
Happy to add it back if it is a critical protection.

**Comment by ranadheerg** [comment]
Hi @jacobtylerwalls After reducing the scope to just nested for loops without the next() call and fixing the for-else bug. I see that the false positives are not there in the primer run.  I'm seeing one looping-through-iterator hit in **ansible** from my local primer run, but it doesn't appear in CI. I confirmed this is not a stale cache issue (reproduced after a fresh clone). This may be due to a difference in the ansible commit being checked locally vs. the pinned version used in CI. Could you take a look at whether this is a genuine bug in ansible or a false positive we need to handle?

https://github.com/ansible/ansible/blob/80ee6b5d920d181fa46b6fdc5e40dca73a2166cc/test/support/integration/plugins/modules/pkgng.py#L316

```
annotate_c = 0
    annotations = map(lambda _annotation:
                      re.match(r'(?P<operation>[\+-:])(?P<tag>\w+)(=(?P<value>\w+))?',
                               _annotation).groupdict(),
                      re.split(r',', annotation))

    operation = {
        '+': annotation_add,
        '-': annotation_delete,
        ':': annotation_modify
    }

    for package in packages:
        for _annotation in annotations:
            if operation[_annotation['operation']](module, pkgng_path, package, _annotation['tag'], _annotation['value']):
                annotate_c += 1

    if annotate_c > 0:
        return (True, "added %s annotations." % annotate_c)
    return (False, "changed no annotations")
```
    
Here the if condition is outside the for loop and it follows the nested for loop iterator producing map function. It satisfies the original issue criteria. annotations is an iterator declared in outer scope and it is called in nested for loop

**Comment by ranadheerg** [comment]
Hi @jacobtylerwalls I've also added the documentation examples. And cleared all the failed checks. Ready for one more review

**Comment by ranadheerg** [comment]
@jacobtylerwalls @Pierre-Sassoulas 

Just a quick ping to see if we can get this moving. To rephrase the current state:

I reduced the scope to just nested for loops (removing next() and while loop checks) as discussed.

This successfully eliminated the false positives we saw in the primer run for Django and Black.

The for-else bug is fixed, docs are updated, and all tests/CI checks are completely green.

The PR is currently blocked by the older "Changes requested" review. Could you take a fresh look when you have a minute and let me know if there's anything else needed to clear the block and get this merged?

Also let me know if we want to expand it for more iterator producing functions, I have added the basic functions 

map, filter, zip, iter and reversed

**Comment by Pierre-Sassoulas** [comment]
Hey @ranadheerg thank you for your work on this. Do you mind rebasing on main ? The primer seems off.

File: doc/data/messages/l/looping-through-iterator/bad.py [hunk 4]
```diff
@@ -0,0 +1,21 @@
+# 1. Generator expressions are exhausted after one use
```
Comment (Pierre-Sassoulas) [inline doc/data/messages/l/looping-through-iterator/bad.py:1]: I think we can keep only one example for the doc, it's to illustrate the issue and the fix not to show that the checker handles multiple situation :)

File: pylint/checkers/looping_iterator_checker.py [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+        "builtins.iter",
+        "builtins.reversed",
+    }
+
+    def __init__(self, linter: PyLinter) -> None:
+        super().__init__(linter)
+        self._scope_stack: list[dict[str, nodes.NodeNG | str]] = []
+
+    # --- Scope Management ---
+
```
Comment (Pierre-Sassoulas) [inline pylint/checkers/looping_iterator_checker.py:48]: The  @utils.only_required_for_messages("looping-through-iterator") is missing for some visitor function, right ?

File: pylint/checkers/looping_iterator_checker.py [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+from astroid import nodes
+
+from pylint import checkers, interfaces
+from pylint.checkers import utils
+
+if TYPE_CHECKING:
+    from pylint.lint import PyLinter
+
+DefinitionType = nodes.NodeNG | str
```
Comment (Pierre-Sassoulas) [inline pylint/checkers/looping_iterator_checker.py]: Unused ?

```suggestion
```

File: pylint/checkers/looping_iterator_checker.py [hunk 7]
```diff
@@ -0,0 +1,193 @@
...
+
+    options = ()
+
+    KNOWN_ITERATOR_PRODUCING_FUNCTIONS: set[str] = {
+        "builtins.map",
+        "builtins.filter",
+        "builtins.zip",
+        "builtins.iter",
+        "builtins.reversed",
+    }
```
Comment (Pierre-Sassoulas) [inline pylint/checkers/looping_iterator_checker.py]:
```suggestion
    KNOWN_ITERATOR_PRODUCING_FUNCTION_QNAMES: set[str] = {
        "builtins.map",
        "builtins.filter",
        "builtins.zip",
        "builtins.iter",
        "builtins.reversed",
    }
```

File: pylint/checkers/looping_iterator_checker.py [hunk 8]
```diff
@@ -0,0 +1,193 @@
...
+            if isinstance(target, nodes.AssignName):
+                variable_name = target.name
+                if is_iterator_definition:
+                    current_scope[variable_name] = value_node
+                else:
+                    current_scope[variable_name] = "SAFE"
+
+    @utils.only_required_for_messages("looping-through-iterator")
+    def visit_call(self, node: nodes.Call) -> None:
+        if isinstance(node.func, nodes.Name) and node.func.name == "next":
```
Comment (Pierre-Sassoulas) [inline pylint/checkers/looping_iterator_checker.py:110]:
```suggestion
        if isinstance(node.func, nodes.Name) and node.func.name == "next":
            # When the user use 'next' it's easy to raise a false positive 
            # and the user usually know what they do so the primer looked 
            # especially bad.
```

**Review by Pierre-Sassoulas — CHANGES_REQUESTED** [review]
This look great overall, I have a few minor comments.

File: pylint/checkers/looping_iterator_checker.py (hunk 3 above)
Comment (jacobtylerwalls) [reply pylint/checkers/looping_iterator_checker.py]: Got it.

Can you simplify this and just model it on astroid's [_get_if_statement_ancestor](https://github.com/pylint-dev/astroid/blob/88d20dc6037d9af5283be89bfc9f1a65e93213ad/astroid/filter_statements.py#L42)()?

File: pylint/checkers/looping_iterator_checker.py [hunk 9]
```diff
@@ -0,0 +1,200 @@
...
+            if isinstance(current, nodes.For):
+                return current
+            if isinstance(current, (nodes.FunctionDef, nodes.ClassDef, nodes.Module)):
+                return None
+            current = current.parent
+        return None  # pragma: no cover — AST always rooted at Module
+
+
+def register(linter: PyLinter) -> None:
+    """This required function is called by Pylint to register the checker."""
```
Comment (jacobtylerwalls) [inline pylint/checkers/looping_iterator_checker.py]:
```suggestion
```

File: tests/functional/l/looping_iterator.py [hunk 10]
```diff
@@ -0,0 +1,420 @@
+"""Functional tests for looping-through-iterator checker."""
```
Comment (jacobtylerwalls) [inline tests/functional/l/looping_iterator.py:1]: This file is missing a test for a third nesting level:

```py
for ...
    for ...
        for ...
```

File: pylint/checkers/looping_iterator_checker.py [hunk 11]
```diff
@@ -0,0 +1,200 @@
...
+            return
+
+        # 3. Get the loop that directly contains the DEFINITION.
+        definition_loop = self._find_ancestor_loop(definition)
+
+        if definition_loop in ancestor_loops_of_usage:
+            return
+
+        inner_loop = ancestor_loops_of_usage[0]
+        outer_loop = ancestor_loops_of_usage[1]
```
Comment (jacobtylerwalls) [inline pylint/checkers/looping_iterator_checker.py]: What happens when one of the loops we care about is at index 2?

**Review by jacobtylerwalls — CHANGES_REQUESTED** [review]

File: pylint/checkers/looping_iterator_checker.py (hunk 9 above)
Comment (Pierre-Sassoulas) [reply pylint/checkers/looping_iterator_checker.py]: Oops, missed that dumb llm comment :sweat: Sorry !

**Comment by Pierre-Sassoulas** [comment]
Pushed a commit addressing the latest review round (thanks @ranadheerg for the groundwork):

**@jacobtylerwalls**
- *Simplify `_find_ancestor_loop` à la `_get_if_statement_ancestor`*: replaced the manual `while`/`.parent` walk with `_enclosing_for_loops()`, which iterates `node.node_ancestors()` and collects the enclosing `for` loops, stopping at the first scope boundary (so a loop in an outer function/class scope is never counted as nesting the usage).
- *"What happens when one of the loops is at index 2?"*: the immediate enclosing loop (`[1]`) is the **sole** determinant of the reuse bug — if it re-runs, the iterator is exhausted on the second pass regardless of any further outer loops. Redefinition at *any* outer level (incl. index 2) is handled separately by the `definition_loop` scan, which now walks every ancestor loop, not just the closest two. The exit-guard correctly keys off `[1]` only.
- *Missing third-nesting-level test*: added four functional tests pinning this down:
  - `three_level_nesting_warns` — consumed in the innermost of three loops → warns
  - `three_level_nesting_exit_in_grandparent_still_warns` — unconditional `return` in the grandparent does **not** save the middle loop → still warns (proves we don't over-suppress on the index-2 path)
  - `three_level_nesting_exit_in_parent_is_safe` — exit in the immediate parent after the inner loop → safe
  - `three_level_nesting_redefined_in_grandparent_is_safe` — redefinition in the grandparent → safe
- Dropped the stray LLM comment in the exit guard.

**@Pierre-Sassoulas**
- `good.py` reduced to a single illustration.
- `leave_for` now also carries `@only_required_for_messages` so the scope stack stays balanced with the decorated `visit_for`.
- (Earlier rounds: `register` docstring removed, `next()` comment, `*_QNAMES` rename, unused alias — all in.)

All hooks + functional tests green.

**Comment by ranadheerg** [comment]
Hi Everyone, Apologies for the delay in response. I had issues syncing the branch with upstream and got carried away with other work priorities. I am not sure if this branch can be merged directly of it has any extra commits that needs cleanup. Thank you @Pierre-Sassoulas  and @jacobtylerwalls for making the changes. Let me know if anything is needed from me to close this out. Enjoyed working on this

**Comment by Pierre-Sassoulas** [comment]
No problem @ranadheerg, I took over because I didn't know if you were going to come back to it, I'm going to take Jacob's remark Into account and merge, thank you for the initial work.

File: doc/data/messages/l/looping-through-iterator/bad.py [hunk 12]
```diff
@@ -0,0 +1,5 @@
...
```
Comment (Pierre-Sassoulas) [inline doc/data/messages/l/looping-through-iterator/bad.py:1]:
```suggestion
```

File: doc/data/messages/l/looping-through-iterator/good.py [hunk 13]
```diff
@@ -0,0 +1,6 @@
+# Materialize the iterator into a collection so it can be reused
```
Comment (Pierre-Sassoulas) [inline doc/data/messages/l/looping-through-iterator/good.py:1]:
```suggestion
```
